rfcs/convergence/native-element-props.md

```diff
@@ -0,0 +1,306 @@
+# RFC: Customizing application of native element props and `ref`
+
+@ecraig12345 @sopranopillow @behowell
+
+## Summary
+
+For certain components, it's necessary to apply top-level native props to an element besides the actual DOM root. This RFC proposes two possible mechanisms for handling this.
+
+Whichever option is chosen, it should be used very sparingly to avoid confusion. The most obvious case is for input components (options for exact criteria discussed below).
+
+## Problem statement
+
+For most components, it makes sense to apply native props (and `ref`) to the root element. Currently the `root` slot always gets native element props passed to the component, and is not explicitly exposed via the API.
+
+The input components are a little unique because the actual `<input>` _cannot_ be the root element of the component, since we need various wrappers for styling (and `<input>` can't have children). However, users may expect that some or all native props passed to the root of the component would be applied to the `<input>` itself. This may also be a requirement for certain 3rd-party form libraries.
+
+For example, rendering a `<Checkbox/>` actually does this _(before the RFC is implemented)_:
+
+```tsx
+<slots.root>
+  <slots.checkmark {/* visual checkmark */} />
+  <slots.input {/* actual input */}>
+  {children /*label*/}
+</slots.root>
+```
+
+Or an `<Input/>` is roughly like this _(again pre-RFC)_:
+
+```tsx
+<slots.root>
+  <slots.bookendBefore />
+  <slots.inputWrapper {/* visually styled as the input */}>
+    <slots.insideStart />
+    <slots.input /> {/* actual input element */}
+    <slots.insideEnd />
+  </slots.inputWrapper>
+  <slots.bookendAfter />
+</slots.root>
+```
+
+So to specify a value, you'd have to use slot props:
+
+```tsx
+<Input input={{ id: 'foo', value: 'stuff' }} />
+```
+
+The question of where to apply the top-level `ref` has similar considerations and should likely follow the same pattern as native props.
+
+### Principles for any solution
+
+The chosen solution needs to adhere to the following principles (thanks @levithomason):
+
+- User needs to retain access to any part of DOM at any time (can't make a piece of DOM that's unable to receive props)
+- Consider what the user expects when using the component
```

Comment:
One way that I've used Fluent components in several applications is along with StyledComponents and Emotion.  I extended or modified the style of a Fluent component using the style extension mechanism.

```tsx
const StyledLink = styled(Link)`
  color: palevioletred;
  font-weight: bold;
`;
```

My expectation is that the styles apply to the root element of the component.  In the checkbox example, If I applied a margin of 4px and this was applied to the input - moving the bookends away from the input part - I would be very surprised.  

When using StyledComponents with Fluent, I often examine the output DOM of a component and then apply selectors to update the style of child parts.  Sometimes this doesn't work because the default part styles override.  I then either need to create a styles object to style that part directly, or apply important to my style.

Whichever solution is chosen, I think it should support the expected behavior of developers working with other styling libraries.